### ISSUE TITLE
10. Cluster_hierarchy

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#51]
"""
비지도 학습 -척도 : 등간, 비율척도(연속적인 양)
K-means : 비계층적 군집분석  
 - 군집의 갯수 알고 함, 확인적
Hierarchical : 계층적 군집분석 
 - 군집의 갯수 모름, 탐색적
 - 처음에 센터로이드 k개가 랜덤하게 설정되고 군집을 형성

### 계층적 군비분석
 - 상향식(Bottom_up)으로 계층적 군집형성
 - 유클리드안 거리계산식 이용
 - 숫자형 변수만 사용 가능
### 군집화 방식
 - 단일기준 결합방식 : 각 군집에서 중심으로부터 거리가 가까운 것(2,3,6) 1개씩 비교하여
 가장 가까운 것 끼리 군집화
 - 완전기준 결합방식 : 각 군집에서 중심으로부터 가장 먼대상(1,4,5) 끼리 비교하여 가장 가
까운 것 끼리 군집화
 - 평균기준 결합방식 : 한 군집 안에 속해 있는 모든 대상과 다른 군집에 속해있는 모든 대상의 쌍 집합에 대한
거리를 평균 계산하여 가장 가까운 것 끼리 군집화
"""

import pandas as pd # dataframe
from sklearn.datasets import load_iris # dataset
from scipy.cluster.hierarchy import linkage, dendrogram # tool
import matplotlib.pyplot as plt # 산점도

### 1. dataset load
iris = load_iris()
iris.feature_names

X=iris.data # 숫자변수
y=iris.target # 숫자변수

-데이터 프레임으로 바꾸기
iris_df = pd.DataFrame(X, columns = iris.feature_names)
sp = pd.Series(y)
-y변수 추가
iris_df['Species'] = sp

iris_df.info()
'''

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
    Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   Species            150 non-null    int32  
dtypes: float64(4), int32(1)
'''

### 2. 계층적 군집분석
-5개 모두 숫자가 되었으므로 군집분석 할 수 있다.
clusters = linkage(iris_df, method='complete', metric='euclidean')
'''
method='single' : 단순연결
method='complete' : 군집화 방식 - 완전기준결합방식
method='average' : 평균연결
metric='euclidean' :유클리디안 거리계산
'''

clusters.shape # (149, 4) # 150-1, 5-1로 거리 계산된 것

### 3. 덴드로그램 시각화
plt.figure(figsize = (10,5)) # 이미지 사이즈
dendrogram(clusters, leaf_rotation = 90, leaf_font_size =20)
plt.show()

### 4. 클러스터 자르기/평가 : 덴드로그램으로 판단
from scipy.cluster.hierarchy import fcluster # cluster 자르기

- 1) 클러스터 자르기
cluster = fcluster(clusters, t=3, criterion ='distance') #  거리를 기준으로 3개의 클러스터로 자르겠다.
cluster # 150개의 관측치를 1~3으로 나타냄

- 2) DF 칼럼추가
iris_df['cluster'] = cluster
iris_df
iris_df.head()# 처음은 1번 군집이 많음
iris_df.tail()# 뒷단은 3번 군집이 많음

- 3) 산점도 시각화
plt.scatter(x = iris_df['sepal length (cm)'], y = iris_df['petal length (cm)'], 
            c = iris_df['cluster'], marker = 'o')
plt.show()

- 4) 관측치 vs 예측치
tab = pd.crosstab(index = iris_df['Species'], columns=iris_df['cluster'])
tab
'''
cluster   1   2   3
Species            
0        50   0   0
1         0   0  50
2         0  34  16
'''

- 5) 군집별 특성분석
-DF.groupby('집단변수') 

cluster_grp = iris_df.groupby('cluster')
cluster_grp.size()
'''
cluster
1    50
2    34
3    66
dtype: int64
'''

cluster_grp.mean()
'''
         sepal length (cm)  sepal width (cm)  ...  petal width (cm)   Species
cluster                                       ...                            
1                 5.006000          3.428000  ...          0.246000  0.000000
2                 6.888235          3.100000  ...          2.123529  2.000000
3                 5.939394          2.754545  ...          1.445455  1.242424

[3 rows x 5 columns]
'''
